Rewrite to be more precompilable.  API BREAKAGE!

staticfloat opened the pull request:
@stevengj I took the opportunity to rewrite basically the whole thing in a much more precompile-friendly manner. On this branch, the time of `julia -e 'using Nettle'` looks like:

|  | `master` | `sf/breakageahoy`
| ----:| ----------|---------------------------
| Julia 0.3 | 1.721s | 1.456s
| Julia 0.4 | 1.436s | 0.554s

If I actually hash something (`julia -e 'using Nettle; md5_hash("a")'` on `master`, `julia -e 'using Nettle; hash!("md5", "a")'` on this branch), there is a difference:

|  | `master` | `sf/breakageahoy`
| ----:| ----------|---------------------------
| Julia 0.3 | 1.726s | 1.560s
| Julia 0.4 | 1.457s | 0.642s

I am a little disappointed that `hash!()` doesn't seem to be getting precompiled; I tried using `SnoopCompile.jl`, and have included the results of my efforts here, but I don't seem to be getting what I want, `@time` output still looks an awful lot like what I would expect from slamming into codegen on first run:

```
$ time ./julia -e 'using Nettle; @time hash!("md5", "a"); @time hash!("md5", "a")'
  0.087612 seconds (70.92 k allocations: 3.103 MB)
  0.000007 seconds (9 allocations: 544 bytes)

real    0m1.101s
user    0m1.094s
sys     0m0.111s
```

@timholy do you have any tips for how to better get this initial runtime down?  For the record, I have done the following:

```julia
using SnoopCompile
using Nettle
@snoop1 "test.csv" hash!("md5", "a")
data = SnoopCompile.read("test.csv")
pc, discards = SnoopCompile.parcel(data[end:-1:1,2])
SnoopCompile.write("/tmp/precompile", pc)
```

Then when I `cat /tmp/precompile/precompile_Nettle.jl`, I get:

```julia
function _precompile_()
    ccall(:jl_generating_output, Cint, ()) == 1 || return nothing
    precompile(Nettle.call, (Type{Nettle.NettleHashType}, ASCIIString, UInt32, UInt32, UIn
t32, Ptr{Void}, Ptr{Void}, Ptr{Void}, Ptr{Void},))
end
```

This was after running `SnoopCompile` on the entire `include(Pkg.dir("Nettle", "test", "runtests.jl"))` thing, since this time around I wanted to see if anything had been missed.  Since it only came up with a single file, I figured that must be the only line I needed to add to my `precompile.jl`, but things just don't seem to be working out like I expected them to.

staticfloat commented:
On a completely unrelated note; this PR completely breaks the API, since we don't want to be autogenerating function names.  I'm pretty happy with what I have here, with the exception of `hash!()`, which is probably going to be the most-called function!

The rest of the similarly-named functions are called things like `hmac()` or `encrypt()`, but because `hash` is both a noun and a verb, I find it very likely that users will want to use a variable called `hash`, and won't appreciate the fact that Nettle's `hash` conflicts with their usage, hence, `hash!()`.  Thoughts?  Should I:

* Change the other convenience functions to `hmac!()` and `encrypt!()`, even though we're not really mutating input variables?

* Change `hash!()` to `hash()` and let the users deal with it?  (They can always access it via `Nettle.hash()` I suppose)

* Change the name to something ugly like `calc_hash()`?

stevengj commented on src/Nettle.jl:
```diff
@@ -23,10 +26,11 @@ end
-hexdigest!(state::Union(HMACState,HashState)) = bytes2hex(digest!(state))
+if VERSION >= v"0.4.0-dev+6521"
+    include("precompile.jl")
+    _precompile_()
```
You should put the `__precompile__()` call at the top of the file, before the `module` statement, to avoid double-parsing the module when it is precompiled.

Note that `precompile()` is supported in 0.3, so it doesn't need to be in a `VERSION` check.

staticfloat replied:
This is the `SnoopCompile` generated stuff; the actual `__precompile__()` stuff is at the top of this file.

stevengj replied:
Oh, I didn't notice the number of underscores.

stevengj commented on src/cipher.jl:
```diff
@@ -33,120 +18,146 @@ immutable NettleCipher
-        result = Array(UInt8,length(source))
-        decrypt!(state,result,source)
-        result
+function EncryptionAlgorithm(name::AbstractString, key)
```
This is capitalized like a type constructor, but it doesn't return a type of the same name.  Wouldn't it be more Julian to actually make this a constructor?

It could be less verbose by calling the function and the type `Encrypter` and `Decrypter`.

staticfloat replied:
Yes, since we're breaking API compatibility anyway, I agree this makes a lot more sense.

stevengj commented on src/hash.jl:
```diff
@@ -2,103 +2,54 @@
-        ccall($fptr_update,Void,(Ptr{Void},Csize_t,Ptr{UInt8}),state.ctx,sizeof(data),pointer(data))
-        state
+# Constructor for NettleHashState
+function HashAlgorithm(name::AbstractString)
```
Again, this looks like a constructor for a type called `HashAlgorithm`, but it returns a `NettleHashState`.  Maybe call the type and the function `Hasher` analogous to my suggestion above for `Encrypter` and `Decrypter`?

stevengj commented on src/hmac.jl:
```diff
@@ -3,35 +3,52 @@
-  write(io, "$(string(T)) HMAC state")
-end
+# Constructor for NettleHMACState
+function HMACAlgorithm(name::AbstractString, key)
```
Same as above: function name does not match type.

staticfloat commented:
@stevengj I have addressed all your comments and decided to go the `calc_hash()` and `calc_hmac()` routes, especially since otherwise `Base.hash()` conflicts with `Nettle.hash()`.

stevengj commented:
Is there any reason why most of the operations are limited to `ASCIIString`s?  I would think you would want to support `ByteString` and `Vector{UInt8}` at least, and maybe `Vector{T}` for any bitstype `T`.

stevengj commented on src/precompile.jl:
```diff
@@ -0,0 +1,31 @@
+# This file autogenerated through https://github.com/timholy/SnoopCompile.jl.  Precompile every day!
+function _precompile_()
+    ccall(:jl_generating_output, Cint, ()) == 1 || return nothing
+    precompile(Nettle.call, (Type{Nettle.Decryptor}, ASCIIString, Array{UInt8, 1},))
```
Why is this Nettle.call and not Base.call, or just call?

stevengj replied:
Oh, I guess `Nettle.call` is needed for 0.3 compat.

stevengj commented on src/cipher.jl:
```diff
@@ -33,120 +17,146 @@ immutable NettleCipher
-        result
+    state = Array(UInt8, cipher_type.context_size)
+    if nettle_major_version >= 3
+        ccall( cipher_type.set_encrypt_key, Void, (Ptr{Void}, Ptr{UInt8}), state, pointer(key))
```
Does this assume that the `key` is NUL-terminated, since I don't see a length being passed?  Maybe use `Cstring` instead of `Ptr{UInt8}` as the argument type then?

staticfloat replied:
Keys are of a specific length.  The key size is [given in the `CipherType` type](https://github.com/staticfloat/Nettle.jl/blob/ae16302796a5e3dee7250ff7828d32fb3f01e20c/src/cipher.jl#L25), and incorrect key sizes [are rejected](https://github.com/staticfloat/Nettle.jl/blob/ae16302796a5e3dee7250ff7828d32fb3f01e20c/src/cipher.jl#L83-L85) by `Encryptor`/`Decryptor`

stevengj commented:
Instead of `calc_hash`, why not just define:
```jl
digest!(state::Hasher, data) = digest!(update!(Hasher, data))
digest!(hashname::AbstractString, data) = digest!(Hasher(hashname), data)
```

staticfloat commented:
That is such a straightforward idea, I'm flabbergasted I didn't think of it before.

staticfloat commented:
I think this is ready now.  @stevengj, would you like me to hold off on merging this into `master` until you've had a chance to get things ready on the IJulia side of things, or should I fire at will?

stevengj commented:
Fire at will.

stevengj commented:
Maybe file issues with Faker.jl, GnuTLS.jl, OAuth.jl, WebSockets.jl, and Yelp.jl to let them know of the API changes?

staticfloat commented:
Thanks, that list was helpful.  Where did you get it?  Can pkg.julialang.org
do reverse-depends or something?
-E

On Mon, Sep 21, 2015 at 1:38 PM, Steven G. Johnson <notifications@github.com
> wrote:

> Maybe file issues with Faker.jl, GnuTLS.jl, OAuth.jl, WebSockets.jl, and
> Yelp.jl to let them know of the API changes?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/staticfloat/Nettle.jl/pull/52#issuecomment-142101894>.
>

stevengj commented:
No, I run the following shell script to fetch all registered Julia packages, then I just `grep Nettle */REQUIRE`:
```
#!/bin/sh

for f in ~/.julia/v0.5/METADATA/*/url; do
    p=$(basename $(dirname $f))
    url=`cat $f`
    echo "updating $p from $url..."
    if test ! -d $p; then
        git clone $url $p
    else
        (cd $p; git pull origin master)
    fi
done
```

stevengj commented:
I've tagged a new version of IJulia that is compatible with both the current Nettle version and the previous version, so it will be fine whenever you tag a new Nettle.
